### PR TITLE
Add `new_from_u64` to `IsaacRng` and `Isaac64Rng`

### PR DESCRIPTION
**Pull request by pitdicker**
This PR includes some clean-up around the init code of ISAAC, support for initializing from an u64, benchmarks for RNG initialization, and some cleanup of the tests.

I have implemented `new_from_u64` for `IsaacRng` and `Isaac64Rng` as a function instead of a trait for now, as that makes it easy to add it to every RNG one at a time.

### CODE REVIEW

**Comment by pitdicker** [comment]
Rebased.

I also removed the `Copy` trait. This turned out to a bit difficult, because initialisation relied on copying an empty static struct. Also `Clone` relied on `Copy`. Initialisation was a bit strange anyway. It followed the same strategy as the C implementation, that expects the initialisation key to be results array of an empty struct. I now made `init` take the key as an argument, and have it create the empty struct itself.

This does move the code around a lot. This makes reviewing more difficult, my apologies. If it helps, it still passes all test and matches the reference implementation (tested).

Once `new_from_u64` is part of `SeedableRng` everything that has to do with initialisation should be in one part of the file.

**Comment by pitdicker** [comment]
One thing I am thinking about is the advise of Bob Jenkins to only use a seed that is half the size of the state. It should be more than enough, but he did not make the change because the reference implementation was already published.

We can leave `from_seed` as it is, so it can match the reference implementation. But for `from_rng` it should not be observable whether we used a seed of only half the state. It should be a free performance win.

File: src/prng/isaac.rs [hunk 1]
```diff
@@ -97,82 +95,82 @@ pub struct IsaacRng {
...
-                a ^= b << 11; d += a; b += c;
-                b ^= c >> 2;  e += b; c += d;
-                c ^= d << 8;  f += c; d += e;
-                d ^= e >> 16; g += d; e += f;
-                e ^= f << 10; h += e; f += g;
-                f ^= g >> 4;  a += f; g += h;
-                g ^= h << 8;  b += g; h += a;
-                h ^= a >> 9;  c += h; a += b;
-            }}
+// Cannot be derived because [u32; 256] does not implement Clone
```
Comment (dhardy) [inline src/prng/isaac.rs]: I think this should be fixed once RFC 2000 gets implemented, but for now this makes sense

File: src/prng/isaac.rs [hunk 2]
```diff
@@ -97,82 +95,82 @@ pub struct IsaacRng {
...
+            rsl: [w(0); RAND_SIZE],
+            mem: mem,
+            a: w(0),
+            b: w(0),
+            c: w(0),
+            cnt: 0,
+        };
+
+        // Prepare the first set of results
+        rng.isaac();
```
Comment (dhardy) [inline src/prng/isaac.rs]: `cnt == 0` so `isaac()` should be run anyway when `next_u32()` is called. Or is this done to try to match results when `seed == 0` or to improve initialisation quality? That's not what the comment says though.

File: src/prng/isaac.rs [hunk 3]
```diff
@@ -272,36 +263,108 @@ impl Rng for IsaacRng {
...
         ::rand_core::impls::next_u128_via_u64(self)
     }
-    
+
     fn try_fill(&mut self, dest: &mut [u8]) -> Result<()> {
         ::rand_core::impls::try_fill_via_u32(self, dest)
     }
 }
 
+/// Creates a new ISAAC-64 random number generator.
```
Comment (dhardy) [inline src/prng/isaac.rs]: This isn't the 64-bit variant, right? Did you copy + paste?

File: src/prng/isaac.rs [hunk 4]
```diff
@@ -272,36 +263,108 @@ impl Rng for IsaacRng {
...
+/// of 0 padding). You do still need to call randinit() to make sure the initial
+/// state isn't uniform-looking."
+/// "After publishing ISAAC, I wanted to limit the key to half the size of r[],
+/// and repeat it twice. That would have made it hard to provide a key that sets
+/// the whole internal state to anything convenient. But I'd already published
+/// it."
+///
+/// And his answer to the question "For my code, would repeating the key over
+/// and over to fill 256 integers be a better solution than zero-filling, or
+/// would they essentially be the same?":
```
Comment (dhardy) [inline src/prng/isaac.rs:265]: Does this have implications for `seed_from_u64` — i.e. adjust all the state instead of just the first bit? (I have no idea myself; maybe it's not significant.)

File: src/prng/isaac.rs [hunk 5]
```diff
@@ -272,36 +263,108 @@ impl Rng for IsaacRng {
...
+    let mut c = w(0xc3163e4b);
+    let mut d = w(0x0f421ad8);
+    let mut e = w(0xd92a4a78);
+    let mut f = w(0xa51a3c49);
+    let mut g = w(0xc4efea1b);
+    let mut h = w(0x30609119);
+
+    let mut mem = [w(0); RAND_SIZE];
+
+    macro_rules! memloop {
```
Comment (dhardy) [inline src/prng/isaac.rs]: Is there a real reason for this obfuscated code (generic wrt source array) — shouldn't it be the same to just use `key` directly (i.e. change fn to `init(mut mem: [w32; RAND_SIZE])`)?

File: src/prng/isaac.rs [hunk 6]
```diff
@@ -97,82 +95,82 @@ pub struct IsaacRng {
...
-                self.mem[i  ] = a; self.mem[i+1] = b;
-                self.mem[i+2] = c; self.mem[i+3] = d;
-                self.mem[i+4] = e; self.mem[i+5] = f;
-                self.mem[i+6] = g; self.mem[i+7] = h;
-            }
+impl IsaacRng {
+    /// Creates an ISAAC random number generator using an u64 as seed.
+    /// If `seed == 0` this will produce the same stream of random numbers as
+    /// the reference implementation when used unseeded.
+    pub fn new_from_u64(seed: u64) -> IsaacRng {
```
Comment (dhardy) [inline src/prng/isaac.rs:123]: I get why you don't want to share code between `init` and `new_from_u64` but it might be worth looking into at some point — well, maybe better to leave it until another PR.

**Review by dhardy — CHANGES_REQUESTED** [review]
Great work! Sorry to be so picky, I know some of what I've commented on isn't your code anyway.

Main things I'm concerned about are the obvious mistake in the comments and potential adjustments to `new_from_u64`. I'm not certain my points are correct, however.

Preferably add new commits rather than rebase now, as this is already complex.

File: src/prng/isaac.rs (hunk 2 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: Yes true. There is no reason to do this step except that it mirrors the reference implementation. It does not matter much whether we pay the cost of the first round during initialization or on the first `next_u32`.

File: src/prng/isaac.rs (hunk 3 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: Yes, a copy + paste error

File: src/prng/isaac.rs (hunk 4 above)
Comment (pitdicker) [reply src/prng/isaac.rs:265]: I don't know all the reasoning behind the initialization code. But with the authors comment

>If the seed is under 32 bytes, they're essentially the same, otherwise repeating the seed would be stronger

I think `seed_from_u64` is fine.

File: src/prng/isaac.rs (hunk 5 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: With the old code it was not possible, but maybe now with a separate key it is. I am going for the simple solution: just try it, and if the result is the same, I will remove the macro. That would be a good improvement!

File: src/prng/isaac.rs (hunk 1 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: I will add a note

File: src/prng/isaac.rs (hunk 6 above)
Comment (pitdicker) [reply src/prng/isaac.rs:123]: It is not that I don't want to share code. Every time you duplicate something there is an extra place to make mistakes. But it took a long read for me to understand how a seed ran trough the old code. The code path is different in 3 separate places, and I didn't find a nice readable way to combine them. I will think about it some more.

**Comment by pitdicker** [comment]
>Sorry to be so picky

Thanks for looking so closely!

> Preferably add new commits rather than rebase now, as this is already complex.

Completely understand. Otherwise you almost have to start again.

File: src/prng/isaac.rs (hunk 2 above)
Comment (dhardy) [reply src/prng/isaac.rs]: It's not about where the cost is paid, it's that this gets run twice. If that's by design, that's fine, but the comment is wrong.

File: src/prng/isaac.rs (hunk 4 above)
Comment (dhardy) [reply src/prng/isaac.rs:265]: Darn, bits != bytes! Sorry.

File: src/prng/isaac.rs (hunk 2 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: O, that would not be by design. After the first round `cnt == RAND_SIZE`. Where would it run for the second time?

Comment (pitdicker) [reply src/prng/isaac.rs]: Note there is one version on the authors site that calls `isaac()` twice, the one used for the [challenge](http://burtleburtle.net/bob/c/rand.c). But the [normal version](http://burtleburtle.net/bob/c/rand.h) only calls it once.

Comment (dhardy) [reply src/prng/isaac.rs]: Ok, you're right — `isaac()` modifies `cnt` of course.

File: src/prng/isaac.rs (hunk 6 above)
Comment (pitdicker) [reply src/prng/isaac.rs:123]: It worked :-). Much better now.

File: src/prng/isaac.rs [hunk 7]
```diff
@@ -115,62 +116,21 @@ impl fmt::Debug for IsaacRng {
...
-        for i in (0..RAND_SIZE/8).map(|i| i * 8) {
-            mix(&mut a, &mut b, &mut c, &mut d, &mut e, &mut f, &mut g, &mut h);
-            mem[i  ] = a; mem[i+1] = b;
-            mem[i+2] = c; mem[i+3] = d;
-            mem[i+4] = e; mem[i+5] = f;
-            mem[i+6] = g; mem[i+7] = h;
-        }
+        let mut key = [w(0); RAND_SIZE];
+        key[0] = w(seed as u32);
+        key[1] += w((seed >> 32) as u32);
```
Comment (dhardy) [inline src/prng/isaac.rs]: `=` then `+=`

Doesn't matter I guess but preferably be consistent

File: src/prng/isaac.rs [hunk 8]
```diff
@@ -115,62 +116,21 @@ impl fmt::Debug for IsaacRng {
...
+        // Initialize with only one pass.
         // A second pass does not improve the quality here, because all of
         // the seed was already available in the first round.
```
Comment (dhardy) [inline src/prng/isaac.rs]: I don't really get this — I mean, all of the seed gets used in the first round when using big keys too, right?

In reality I suppose it shouldn't matter since a 64-bit seed is not really sufficient for cryptography and this method should still be sufficient for non-crypto uses I guess (unless there's any detectable bias in initial output). Speaking very much as a non-expert here...

**Review by dhardy — COMMENTED** [review]
Definitely like the new init code

File: src/prng/isaac.rs (hunk 8 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: For a short overview: in blocks of 8 words the seed is mixed with the temporary array `a`...`h`. Each block of 8 words then becomes part of the new state. After one full pass every word of the new state is effected by the first block. But only the last block of 8 words is effected by the last block of the seed.

After the first pass `a`...`h` has all of the seed mixed into it. Doing another pass will now mix bits from the middle and last blocks of the seed into the entire state.

The 64-bit seed is part of the first block, and therefore available to `a`...`h` from the first round. At the end of the first full pass the 'random quality' of `a`...`h` is not any higher than at the start. A second pass does not distribute the randomness from the seed more evenly than the first.

File: src/prng/isaac.rs (hunk 7 above)
Comment (pitdicker) [reply src/prng/isaac.rs]: I should be more careful, this are sloppy mistakes...

**Comment by pitdicker** [comment]
I have tried a merge commit, hope I did it right...

File: src/prng/isaac.rs (hunk 8 above)
Comment (dhardy) [reply src/prng/isaac.rs]: Ah, I get you. But the 32-bit generator has two blocks set; `b` will not affect `a` until the second round if I understand correctly. I don't see why it matters though (provided `a` is initially set to a good seed, which it should be, since it gets added to the golden ratio).
